src/legacy/core_plugins/kibana/public/discover/np_ready/angular/discover_state.ts

```diff
@@ -0,0 +1,244 @@
+  /**
+   * Columns displayed in the table
+   */
+  columns?: string[];
+  /**
+   * Array of applied filters
+   */
+  filters?: Filter[];
+  /**
+   * id of the used index pattern
+   */
+  index?: string;
+  /**
+   * Used interval of the histogram
+   */
+  interval?: string;
+  /**
+   * Lucence or KQL query
+   */
+  query?: Query;
+  /**
+   * Array of the used sorting [[field,direction],...]
+   */
+  sort?: string[][];
+}
+
+interface GetStateParams {
+  /**
+   * Default state used for merging with with URL state to get the initial state
+   */
+  defaultAppState?: AppState;
+  /**
+   * Determins the use of long vs. short/hashed urls
+   */
+  storeInSessionStorage?: boolean;
+  /**
+   * Browser history used for testing
+   */
+  hashHistory?: History;
+}
+
+export interface GetStateReturn {
+  /**
+   * kbnUrlStateStorage
+   */
+  kbnUrlStateStorage: IKbnUrlStateStorage;
+  /**
+   * App state, the _a part of the URL
+   */
+  appStateContainer: ReduxLikeStateContainer<AppState>;
+  /**
+   * Start sync between state and URL
+   */
+  startSync: () => void;
+  /**
+   * Stop sync between state and URL
+   */
+  stopSync: () => void;
+  /**
+   * Set app state to with a partial new app state
+   */
+  setAppState: (newState: Partial<AppState>) => void;
+  /**
+   * Get global filters
+   */
+  getAppFilters: () => Filter[];
+  /**
+   * Sync state to URL, used for testing
+   */
+  flushToUrl: () => void;
+  /**
+   * Reset initial state to the current app state
+   */
+  resetInitialAppState: () => void;
+  /**
+   * Return the Appstate before the current app state, useful for diffing changes
+   */
+  getPreviousAppState: () => AppState;
+  /**
+   * Returns whether the current app state is different to the initial state
+   */
+  isAppStateDirty: () => void;
+  /**
+   * Replace the current URL with the current state without adding another browser history entry
+   */
+  replaceUrlState: (startSync: boolean) => Promise<void>;
+}
+const APP_STATE_URL_KEY = '_a';
+
+/**
+ * Builds and returns appState and globalState containers and helper functions
+ * Used to sync URL with UI state
+ */
+export function getState({
+  defaultAppState = {},
+  storeInSessionStorage = false,
+  hashHistory,
+}: GetStateParams): GetStateReturn {
+  const stateStorage = createKbnUrlStateStorage({
+    useHash: storeInSessionStorage,
+    history: hashHistory ? hashHistory : createHashHistory(),
+  });
+
+  const appStateFromUrl = stateStorage.get(APP_STATE_URL_KEY) as AppState;
+  let initialAppState = {
+    ...defaultAppState,
+    ...appStateFromUrl,
+  };
+  let previousAppState: AppState;
+  const appStateContainer = createStateContainer<AppState>(initialAppState);
+
+  const appStateContainerModified = {
+    ...appStateContainer,
+    ...{
+      set: (value: AppState | null) => {
+        if (value) {
+          previousAppState = appStateContainer.getState();
+          appStateContainer.set(value);
+        }
+      },
+    },
+  };
```

Comment:
Would it look better without additional destructuring?
```
const appStateContainerModified = {
    ...appStateContainer,
    set: (value: AppState | null) => {
        if (value) {
          previousAppState = appStateContainer.getState();
          appStateContainer.set(value);
        }
    },
  };
```

Reply:
so much better! done!